src/cluster_utils/server/utils.py

```diff
@@ -85,6 +86,35 @@ class ClusterRunType(enum.Enum):
     HP_OPTIMIZATION = 1
 
 
+class SignalWatcher:
+    """Watch for a signal."""
+
+    def __init__(self, signal_to_watch_for: int = signal.SIGINT) -> None:
+        """
+        Args:
+            signal_to_watch_for: The signal to watch for.
+        """
+        self.received_signal = False
+        signal.signal(signal_to_watch_for, self._signal_handler)
+
+    def _signal_handler(self, sig, frame) -> None:
```

Comment:
I did a quick test:
```
    watcher = SignalWatcher()
    del watcher
    gc.collect()  # to make sure garbage is collected

    for _ in range(10):
        time.sleep(1)
```
When pressing Ctrl+C during the loop, it doesn't crash and the function of SignalWatcher actually still gets called (I confirmed by adding a print there).  Probably it keeps a reference somewhere for the signal handler and thus it doesn't get garbage collected.
I confirmed this by adding a `__del__` method with a print:  It doesn't get called on `del watcher` but only when registering a new signal handler.

My next thought was to register SIG_DFL in `__del__` just in case, but this is actually bad as this will get called _after_ registering a new handler.  That is
```
    watcher = SignalWatcher()
    del watcher
    signal.signal(signal.SIGINT, my_handler)
    # at this point the SIG_DFL is active, not my_handler
```

So TL;DR: It seems to be save as is, so I wouldn't try to actively do any handler cleanup.

Reply:
I see. Well let's keep it as is.